JDBC: Use PrepareStatement

singhpk234 opened the pull request:
<!--
    Possible security vulnerabilities: STOP here and contact security@apache.org instead!

    Please update the title of the PR with a meaningful message - do not leave it "empty" or "generated"
    Please update this summary field:

    The summary should cover these topics, if applicable:
    * the motivation for the change
    * a description of the status quo, for example the current behavior
    * the desired behavior
    * etc

    PR checklist:
    - Do a self-review of your code before opening a pull request
    - Make sure that there's good test coverage for the changes included in this PR
    - Run tests locally before pushing a PR (./gradlew check)
    - Code should have comments where applicable. Particularly hard-to-understand
      areas deserve good in-line documentation.
    - Include changes and enhancements to the documentation (in site/content/in-dev/unreleased)
    - For Work In Progress Pull Requests, please use the Draft PR feature.

    Make sure to add the information BELOW this comment.
    Everything in this comment will NOT be added to the PR description.
-->

eric-maynard commented on persistence/relational-jdbc/src/main/java/org/apache/polaris/persistence/relational/jdbc/QueryGenerator.java:
```diff
@@ -80,114 +99,123 @@ public static String generateDeleteQueryForEntityPolicyMappingRecords(
 
-  public static String generateDeleteQuery(
+  public static PreparedQuery generateDeleteQuery(
       @Nonnull Class<?> entityClass, @Nonnull Map<String, Object> whereClause) {
```
technically the keys in `whereClause` are vulnerable as well, though these are always passed in from the `toMap` of a model today

dimas-b replied:
Perhaps SQL generation ought to be delegated to each entity class. This way the set of columns can be clearly linked to the compile-time java object fields.

dimas-b commented on persistence/relational-jdbc/src/main/java/org/apache/polaris/persistence/relational/jdbc/QueryGenerator.java:
```diff
@@ -37,35 +40,51 @@
+  }
+
+  public static <T> PreparedQuery generateSelectQuery(
       @Nonnull Converter<T> entity, @Nonnull Map<String, Object> whereClause) {
```
Ideally, SQL generation should be based on object type and use case, but not depend on runtime parameters.

dimas-b replied:
Suggestion: keep SQL scoped to classes (basically `static` data). Also, mind IJ warnings about "unsafe" SQL - I beleive IJ's SQL analysis is fairly good. If it flags and issue, there usually is an issue.

singhpk234 replied:
IJ is flagging this as well : PreparedStatement statement = connection.prepareStatement(query.getSql()) ideally this should not be flagged right since we are using `?` and setting values later ?

singhpk234 replied:
keys are something we decide inside the application and don't take from user input, never the less i also added a check to do a check that if where clause keys are subset of the model key else throw Illegal state exception.
Can you please elaborate on `this then column names will be fixed, no runtime input -> safer SQL` ?

Orthogonal to irrespective of who generates SQL if i feed this in prepare statement to the IJ complains

dimas-b replied:
to clarify: SQL text depends on two factors: operation (list, write, read, etc.) and entity type. The latter is represented by java classes. We should be able to keep a few pameterized SQL strings in those classes and use them depending on the operation. I believe the SQL text can be defined at compile time. The same classes can call `.prepareStatement(String)` to avoid any ambiguity as to the source and content of SQL statements. Parameters will be injected later with the help of classes representing entities - that would be similar to current `toMap()` methods, but the keys/column names would be implicit. Keeping SQL and param injection in the same class ensures that values match placeholders.

adutra replied:
I tend to agree with @dimas-b, SQL should be statically defined at compile time for maximum security and efficiency.

It should be possible to achieve that by declaring a few constants in each model class. e.g. for `ModelEntity`:

```java
public class ModelEntity implements Converter<PolarisBaseEntity> {

  public static final String TABLE_NAME = "ENTITIES";

  public static final List<String> PK_COLUMNS = List.of("realm_id", "id");

  public static final List<String> ALL_COLUMNS =
      List.of(
          "realm_id",
          "id",
          "catalog_id",
          "parent_id",
          "type_code",
          "name",
          "entity_version",
          "sub_type_code",
          "create_timestamp",
          "drop_timestamp",
          "purge_timestamp",
          "to_purge_timestamp",
          "last_update_timestamp",
          "properties",
          "internal_properties",
          "grant_records_version");

  public static final String LOOKUP_QUERY =
      "SELECT "
          + String.join(", ", ALL_COLUMNS)
          + " FROM "
          + TABLE_NAME
          + " WHERE "
          + String.join(" = ? AND ", PK_COLUMNS)
          + " = ?";

  public static final String LIST_QUERY =
      "SELECT "
          + String.join(", ", ALL_COLUMNS)
          + " FROM "
          + TABLE_NAME
          + " WHERE realm_id = ? AND catalog_id = ? AND parent_id = ? AND type_code = ?";

  public static void bindLookupQuery(PreparedStatement stmt, ModelEntity entity, String realmId) throws SQLException {
    stmt.setObject(1, realmId);
    stmt.setObject(2, entity.getId());
  }
  
  public static void bindListQuery(PreparedStatement stmt, ModelEntity entity, String realmId) throws SQLException {
    stmt.setObject(1, realmId);
    stmt.setObject(2, entity.getCatalogId());
    stmt.setObject(3, entity.getParentId());
    stmt.setObject(4, entity.getTypeCode());
  }
...
```

dimas-b replied:
`PreparedStatement` by itself is not very meaningful unless the SQL that goes into it is detached from the runtime parameters. The `PreparedStatement` ensures that parameters are separated at execution time. I believe it is also important to isolate runtime parameters at SQL text generation time (I do not think all SQL in this Persistence can be constant, but we should make the SQL generation work without using runtime entities as input).

singhpk234 replied:
[Update] I refactored the QueryGenerator to now containing only 5 ways to generate the query so now it has only 
1. SELECT and where 
2. UPDATE all columns and where 
3. DELETE and where 
4. SELECT with IN query with dynamic list 
5. SELECT with one special disjunction 

These are all single table queries, all the projection are taken from the static member and the filters column reference has an assertion that keys in filter are subset of the columns from the static member of the model. 
All this functions return a PreparedQuery -> SQL string with `?` and the ordered list of parameters which we can feed to the JDBC

Need to refactor this more to completely decouple the query generation and parameter feeding, working on it.

dimas-b commented on persistence/relational-jdbc/build.gradle.kts:
```diff
@@ -34,6 +34,7 @@ dependencies {
   compileOnly(libs.jakarta.inject.api)
 
   implementation(libs.smallrye.common.annotation) // @Identifier
+  implementation("org.postgresql:postgresql:42.7.3")
```
please use the `toml` file for lib versions

singhpk234 commented on persistence/relational-jdbc/src/main/java/org/apache/polaris/persistence/relational/jdbc/models/ModelEntity.java:
```diff
@@ -178,12 +180,28 @@ public Map<String, Object> toMap() {
-    map.put("properties", this.getProperties());
-    map.put("internal_properties", this.getInternalProperties());
+    if (databaseType.equals(DatabaseType.POSTGRES)) {
+      map.put("properties", toPGobject(this.getProperties()));
```
PG specialization is now required for the CI to make happy as what happens unlinke Statements prepareStatement expects the PG type hence, i added the same.

dimas-b replied:
ok. fair enough.

adutra commented on persistence/relational-jdbc/build.gradle.kts:
```diff
@@ -34,6 +34,7 @@ dependencies {
   compileOnly(libs.jakarta.inject.api)
 
   implementation(libs.smallrye.common.annotation) // @Identifier
+  implementation(libs.postgresql)
```
This is creating a hard dependency from this project to the Postgres JDBC driver, FYI.

dimas-b replied:
@singhpk234 's reply here: https://github.com/apache/polaris/pull/1802#discussion_r2124954152

adutra commented on persistence/relational-jdbc/src/main/java/org/apache/polaris/persistence/relational/jdbc/QueryGenerator.java:
```diff
@@ -205,9 +227,14 @@ public static String getTableName(@Nonnull Class<?> entityClass) {
+  }
 
-    return tableName;
+  private void checkInvalidColumns(Map<String, Object> entity, Map<String, Object> whereClause) {
+    List<String> allColumns = new ArrayList<>(entity.keySet());
+    allColumns.add("realm_id");
+    if (!allColumns.containsAll(whereClause.keySet())) {
+      throw new IllegalArgumentException("Invalid query " + whereClause.keySet());
+    }
   }
```
```suggestion
  private void checkInvalidColumns(Map<String, Object> entity, Map<String, Object> whereClause) {
    if (!whereClause.isEmpty()) {
      for (String key : whereClause.keySet()) {
        if (!entity.containsKey(key) && !key.equals("realm_id")) {
          throw new IllegalArgumentException("Invalid query column: " + key);
        }
      }
    }
```

adutra commented on persistence/relational-jdbc/src/main/java/org/apache/polaris/persistence/relational/jdbc/QueryGenerator.java:
```diff
@@ -34,38 +37,67 @@
-    return generateSelectQuery(entity, generateWhereClause(whereClause));
+
+    String tableName = getTableName(entity.getClass());
+    Map<String, Object> objectMap = entity.toMap(databaseType);
```
This is very inefficient; we don't need to collect the object's current values, we only need the keys (column names).

adutra replied:
... and you are doing this again and again for _every_ SQL query, while doing this only once and storing the result in a static field would be enough. I think this qualifies as "very inefficient".

singhpk234 replied:
I agree with you this being done again and again, where there could be better ways to achieve it and infact i took your recommendation as well, the way i see "very inefficient" is that we might have performance implication of doing it, which IMHO should not be, since they are short lived object with very small memory print.

I totally respect your feedback and have addressed the same, since the concerns are addressed, I am marking this thread as resolved.

adutra commented on persistence/relational-jdbc/src/main/java/org/apache/polaris/persistence/relational/jdbc/JdbcBasePersistenceImpl.java:
```diff
@@ -269,8 +291,7 @@ public PolarisBaseEntity lookupEntity(
         Map.of("catalog_id", catalogId, "id", entityId, "type_code", typeCode, "realm_id", realmId);
-    String query = generateSelectQuery(new ModelEntity(), params);
-    return getPolarisBaseEntity(query);
+    return getPolarisBaseEntity(queryGenerator.generateSelectQuery(new ModelEntity(), params));
```
Not related to this change, but why do we need so many WHERE conditions? The PK is `(realmId, id)`, so that should be sufficient.

singhpk234 commented on persistence/relational-jdbc/src/main/java/org/apache/polaris/persistence/relational/jdbc/QueryGenerator.java:
```diff
@@ -21,193 +21,209 @@
-    return "DELETE FROM " + tableName + whereConditions;
+  @VisibleForTesting
+  PreparedQuery generateSelectQuery(
+      @Nonnull List<String> columnNames, @Nonnull String tableName, @Nonnull String filter) {
```
This method should have been private and is used strictly within QueryGenerator itself, this is made package protected with explicit @VisbleforTesting to test coverage, if this is causing some confusion or un-ease I am happy to remove this method or test for this method

singhpk234 replied:
I changed this to private and removed the UT if brings peace.

adutra replied:
Thanks, please remove the `@VisibleForTesting` annotation as well.

dimas-b commented on persistence/relational-jdbc/src/main/java/org/apache/polaris/persistence/relational/jdbc/JdbcBasePersistenceImpl.java:
```diff
@@ -412,8 +464,10 @@ public <T> Page<T> listEntities(
     try {
+      PreparedQuery query =
+          queryGenerator.generateSelectQuery(
+              ModelEntity.ALL_COLUMNS, ModelEntity.TABLE_NAME, params);
```
I believe it would be preferable to completely isolate SQL generation from input values... E.g. use `params.keySet()` instead of all `params`.

Taking this one step further, since the key set is known for this use case, we could have a specialized generator like `LookupEntitiesQuery.sql()`.

Currently, SQL specializations are contained in methods that call `queryGenerator.generateSelectQuery()`. I'm proposing to make sub-classes for those cases. I believe the amount of code is going to be roughly the same, but we can achieve static SQL generation is most cases. In the remaining cases, I'm sure we can still take runtime entities out of the SQL generation methods (perhaps the only non-static case is going to be `IN` queries). WDYT?

singhpk234 replied:
> I believe it would be preferable to completely isolate SQL generation from input values... E.g. use params.keySet() instead of all params.

I agree with you, but presently, we just use params.keySet even today for query generation the user values are just feed to the PreparedQuery record's parameter, i can think off, specially since we assert the keys in params are strict subset of the columns of the table. checking if i can stream line it more and do the PreparedQuery in JDBC persistence it self

> Taking this one step further, since the key set is known for this use case, we could have a specialized generator like LookupEntitiesQuery.sql().

There are 33 sql's presently, I agree static would be preferable, I am parallely trying to do the same as well, here is a gist in which i have drilled it down into all static queries  late last night https://gist.github.com/singhpk234/a63e7236a570892fcaf46bbf4cb89f34 except IN (?)  but i haven't tested it E2E, would need some more time for that, just putting this pr out to address immediate concerns.

dimas-b commented on persistence/relational-jdbc/src/main/java/org/apache/polaris/persistence/relational/jdbc/DatasourceOperations.java:
```diff
@@ -206,6 +215,17 @@ public void runWithinTransaction(TransactionCallback callback) throws SQLExcepti
 
+  public Integer execute(Connection connection, QueryGenerator.PreparedQuery preparedQuery)
+      throws SQLException {
+    try (PreparedStatement statement = connection.prepareStatement(preparedQuery.sql())) {
```
This code appears to be called one per entity in `writeEntities`. I believe we should try and refactor the code to do `prepareStatement` one, execute many in that case.

dimas-b replied:
I did not mean preparing at startup, but at least reusing the `PreparedStatement` inside loops.

dimas-b commented on persistence/relational-jdbc/src/main/java/org/apache/polaris/persistence/relational/jdbc/QueryGenerator.java:
```diff
@@ -21,193 +21,211 @@
+  public PreparedQuery generateDeleteQueryForEntityGrantRecords(
+      @Nonnull PolarisEntityCore entity, @Nonnull String realmId) {
+    String where =
+        " WHERE ((grantee_id = ? AND grantee_catalog_id = ?) OR (securable_id = ? AND securable_catalog_id = ?)) AND realm_id = ?";
```
nit: maybe use multi-line `"""` strings with shorter lines for ease of reading (in IDE)?

dimas-b commented on persistence/relational-jdbc/src/main/java/org/apache/polaris/persistence/relational/jdbc/models/ModelEntity.java:
```diff
@@ -178,8 +200,13 @@ public Map<String, Object> toMap() {
     map.put("last_update_timestamp", this.getLastUpdateTimestamp());
-    map.put("properties", this.getProperties());
-    map.put("internal_properties", this.getInternalProperties());
+    if (databaseType.equals(DatabaseType.POSTGRES)) {
+      map.put("properties", toJsonbPGobject(this.getProperties()));
```
suggestion: use an interface for converting values to driver-specific types and provide different implementations per `DatabaseType` (it could even be obtained from `DatabaseType.converter()` perhaps) instead of `if`s.

dimas-b reviewed (approved):
As far as I was able to follow the proposed new code, generated SQL appears to be a function of object _types_ and use cases. No runtime entity values leak into SQL text that is used to make `PreparedStatement` instances.

I still think SQL generation can be improved to be more isolated from runtime values (it can be done statically in many cases), but I'd be ok doing that in a follow up PR.

flyrain reviewed (commented):
Sorry for the late review. Thanks a lot for the quick fix, @singhpk234! LGTM overall. Left minor comments.

singhpk234 commented:
Thank you so much every one for the quick review, I tried addressing most of the feedback here : 
Re: Static SQL - I opened a pr https://github.com/apache/polaris/pull/1824 have moved everything to the SQL Constants to make this entirely static 
Re: Realm-id as part of model - I addressed this in https://github.com/apache/polaris/pull/1824 as well now the model will contain the realm-id as the part of model itself, i didn't intially put this as PolarisBaseEntity has no way of realm-id, but more i think of it now and as called out in the PR feedbacks that the right way to go !
Re: DatabaseType - I agree this is leaking rightnow in the model, let me think of a more cleaner way to abstract it, I like @dimas-b  suggestion to make the converter at from the type level, let me take stab in this refactoring. 

I hope this addresses major concern around this PR and i sincerly thank the whole community to jump in and help here !

jbonofre commented:
@singhpk234 awesome, I'm cherry picking on `release/0.10.x` branch and prepare the RC5 😄
